include/kernel/thread.h

```diff
@@ -286,4 +290,59 @@ struct k_thread {
 typedef struct k_thread _thread_t;
 typedef struct k_thread *k_tid_t;
 
+#ifdef CONFIG_SMP
+
+/* True if the current context can be preempted and migrated to
+ * another SMP CPU.
+ */
+bool z_smp_cpu_mobile(void);
+
+/* These values are performance critical and we want the compiler to
+ * be able to optimize their results, even if they are defined in
+ * architecture code that might otherwise preclude that (for example
+ * as C function calls, volatile __asm__ blocks, etc...).
+ */
+
+/* The current CPU may change, but only once we release a lock and
+ * become preemptiable.  This function is "pure", meaning that the
+ * compiler will only trust its resutl until global program state may
+ * have changed.  So the memory clobbers (or just function calls)
+ * required by arch locking operations will force it to be
+ * re-evaluated.
+ */
+static ALWAYS_INLINE __attribute__((pure)) struct _cpu *__eval_curr_cpu(void)
+{
+	return arch_curr_cpu();
+}
+
+/* The current thread is by definition unchanging from the perspective
+ * of generated code, so this is "const" allowing the compiler to
+ * cache its value in any manner it wants.
+ */
+static ALWAYS_INLINE __attribute__((const)) struct k_thread *__eval_curr(void)
+{
+	return __eval_curr_cpu()->current;
```

Comment:
Hmm.

This can result in serious misbehavior if called in a spot that can preempt. And I spot several cases in Zephyr that appears to use _current without an interrupt lock...

Reply:
That's actually the main purpose of `k_current_get`, by the way. It locks interrupts, grabs `_current_cpu->current`, then unlocks interrupts.

Reply:
I assure you (again, based on the assert we were just discussing)  that nowhere in Zephyr reachable by code on a platform with SMP enabled evaluates _current_cpu in a preemptible context.  And any places in source code that you find are bugs we should fix, they simply happen to have slipped through our existing testing.

It's never legal to do that.  And therefore, it's perfectly OK to assume that we are non-preemptible here.  In any case this patch doesn't change the mechanism, which is already in place.  All this is about is optimization.

Reply:
I think we're talking past each other here.

To be clear: *I am **not** talking about `_current_cpu`*. I am talking about *_current*, which this patch changes to use *__current_cpu* internally.

Previously, `_current` was redirected to `k_current_get`, which 1. locks interrupts, 2. grabs `_current_cpu->current`, and 3.  unlocks interrupts.

So previously accesses to `_current` did use `_current_cpu` under the hood, so to speak, but the interrupt lock / unlock ensured that the access was always done with interrupts locked.

With this patch, `_current` directly evaluates to `__eval_curr_cpu()->current`, *which does not lock interrupts*.

What this means is that previous uses of `_current` (Again, note: uses of **_current**. _Not_ uses of `_current_cpu`) were always safe regardless of if we were preemptable. And any such uses now suddenly are **unsafe** if we are preemptable, because it **now* directly calls into `__eval_curr_cpu` **without** an interrupt lock.

(...probably too much emphasis. My apologies; I am trying to clarify here.)

Reply:
Or, to put it another way.

Consider the implementation of, for instance, `k_mbox_put`:

```c
int k_mbox_put(struct k_mbox *mbox, struct k_mbox_msg *tx_msg,
	       k_timeout_t timeout)
{
	/* configure things for a synchronous send, then send the message */
	tx_msg->_syncing_thread = _current;
```
This function is a public `k_` function, and can be called from anywhere, including userspace code. It is **not** documented as requiring a spinlock or interrupt locking.

With the previous implementation, this is safe. It roughly expanded to:

```c
int k_mbox_put(struct k_mbox *mbox, struct k_mbox_msg *tx_msg,
	       k_timeout_t timeout)
{
#ifdef CONFIG_SMP
	unsigned int k = arch_irq_lock();
#endif
	k_tid_t ret = arch_curr_cpu()->current;
#ifdef CONFIG_SMP
	arch_irq_unlock(k);
#endif
	tx_msg->_syncing_thread = ret;
```
...which is safe, because `arch_curr_cpu()`'s result is only used within the irq lock.

**With** this patch, it now expands to, roughly

```c
int k_mbox_put(struct k_mbox *mbox, struct k_mbox_msg *tx_msg,
	       k_timeout_t timeout)
{
	k_tid_t ret = arch_curr_cpu()->current;
	tx_msg->_syncing_thread = ret;
```
**Note the missing lock.** Now consider what happens if we switch contexts between the call to `arch_curr_cpu` and the dereference of the resulting cpu struct's`current`...

Reply:
Alternatively:

> In any case this patch doesn't change the mechanism,

Previously `_current` was redirected through `k_current_get`, which wraps a call to `arch_curr_cpu` and resulting dereference in an irq lock.
Now current is directly a call to `arch_curr_cpu` and resulting dereference **with no irq lock**.

Reply:
Hm. Apparently I'm hungry. My apologies for the above, they are probably a mite snippy. I'm going to stop and go get food...

Reply:
Heh, and I assure you, based on the assert we've been discussing, that there are no uses of _current (not just _current_cpu) reachable by code on a platform with SMP enabled, that happen in a preemptible context.  And, again, that any such spots that appear to exist are bugs to be fixed (and either dead code or in drivers or subsystems that don't target SMP).

It's true that the previous #define to k_current_get() would have hidden such a bug if it existed.  But no such bugs existed.

(and to clarify: _current is an internal thing, no out-of-tree applications are allowed to touch it, they have to go through k_current_get() which is safe).

Reply:
> But no such bugs existed.

You keep on asserting this; this appears to be a seriously incorrect assumption.

Or rather: you have an assumption here that "`_current` can never be called from an unlocked context" which is incorrect. The previous implementation _can_ be called from an unlocked context and works just fine.

Can you please look at my previous example of `k_mbox_put`, which (seemingly) does **precisely** this?

Or, for that matter, the start of arc's `arch_user_mode_enter`, which is called from `k_thread_user_mode_enter` (again: a public API) in an unlocked context and does:

```c
FUNC_NORETURN void arch_user_mode_enter(k_thread_entry_t user_entry,
					void *p1, void *p2, void *p3)
{
	setup_stack_vars(_current);
```

Or, for that matter, the start of _aarch32's_ `arch_user_mode_enter`, which does pretty much the same thing:

```c
FUNC_NORETURN void arch_user_mode_enter(k_thread_entry_t user_entry,
					void *p1, void *p2, void *p3)
{

	/* Set up privileged stack before entering user mode */
	_current->arch.priv_stack_start =
		(uint32_t)z_priv_stack_find(_current->stack_obj);
```

Or the start of aarch32's `arch_float_disable`, which is called from `k_float_disable` (again, a public API!) and does:

```c
int arch_float_disable(struct k_thread *thread)
{
	if (thread != _current) {
		return -EINVAL;
	}
```

Or cortex-m's `z_impl_k_thread_abort`, which is called from `k_thread_abort` in an unlocked context (again, a public API) and does:

```c
void z_impl_k_thread_abort(k_tid_t thread)
{
	if (_current == thread) {
```

Or riscv's z_impl_user_fault, which is called from `k_user_fault` in an unlocked context:

```c
void z_impl_user_fault(unsigned int reason)
{
	z_arch_esf_t *oops_esf = _current->syscall_frame;
```

Or riscv's `arch_float_disable` for that matter:

```c
int arch_float_disable(struct k_thread *thread)
{
	unsigned int key;

	if (thread != _current) {
```

Or, for that matter, `z_setup_new_thread` itself:

```c
#ifdef CONFIG_ARCH_HAS_CUSTOM_SWAP_TO_MAIN
	/* _current may be null if the dummy thread is not used */
	if (!_current) {
```
(which is where the Aarch64 SMP port breaks with this regression by the way.)

I could go on.

> Heh, and I assure you, based on the assert we've been discussing, that there are no uses of _current (not just _current_cpu) reachable by code on a platform with SMP enabled

Adding this change to the AArch64 SMP port breaks it utterly (doesn't boot, see above for the first failing spot).

It may be your intent to change `_current` to only be usable in an unlocked context. But simply declaring that this is not a change does not make it so. The current code _does_ assume that `_current` can be called from an unlocked context, in many places, **and this is perfectly safe to do so in the current implementation**. 

If you wish to change all uses of `_current` to prevent them from being called from an unlocked context - and deal with the ensuing regressions - sure, go ahead I guess.

*****

> (and to clarify: _current is an internal thing, no out-of-tree applications are allowed to touch it, they have to go through k_current_get() which is safe).

Uh, `_current` is used in in-tree drivers. (Notably, eswifi.) Are out-of-tree drivers a second-class citizen then?